src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1022 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This contains the synced kv log store implementation.
+//! It's meant to buffer a large number of records emitted from upstream,
+//! to avoid overwhelming the downstream executor.
+//!
+//! The synced kv log store polls two futures:
+//!
+//! 1. Upstream: upstream message source
+//!
+//!    It will write stream messages to the log store buffer. e.g. `Message::Barrier`, `Message::Chunk`, ...
+//!    When writing a stream chunk, if the log store buffer is full, it will:
+//!      a. Flush the buffer to the log store.
+//!      b. Convert the stream chunk into a reference (`LogStoreBufferItem::Flushed`)
+//!         which can read the corresponding chunks in the log store.
+//!         We will compact adjacent references,
+//!         so it can read multiple chunks if there's a build up.
+//!
+//!    On receiving barriers, it will:
+//!      a. Apply truncation to historical data in the logstore.
+//!      b. Flush and checkpoint the logstore data.
+//!
+//! 2. State store + buffer + recently flushed chunks: the storage components of the logstore.
+//!
+//!    It will read all historical data from the logstore first. This can be done just by
+//!    constructing a state store stream, which will read all data until the latest epoch.
+//!    This is a static snapshot of data.
+//!    For any subsequently flushed chunks, we will read them via
+//!    `flushed_chunk_future`. See the next paragraph below.
+//!
+//!    We will next read `flushed_chunk_future` (if there's one pre-existing one), see below for how
+//!    it's constructed, what it is.
+//!
+//!    Finally we will pop the earliest item in the buffer.
+//!    - If it's a chunk yield it.
+//!    - If it's a watermark yield it.
+//!    - If it's a flushed chunk reference (`LogStoreBufferItem::Flushed`),
+//!      we will read the corresponding chunks in the log store.
+//!      This is done by constructing a `flushed_chunk_future` which will read the log store
+//!      using the `seq_id`.
+//!   - Barrier,
+//!     because they are directly propagated from the upstream when polling it.
```

Comment:
See this docstring for the design.

Reply:
So what does "synced" mean here? (and what is unsynced logstore?)

Reply:
unsynced logstore is our current design for the sink logstore, where we have 2 separate streams for reading from and to the logstore.

Synced means that we have a single stream instead, rather than 2 concurrently running ones.

Reply:
I'm not very familiar with our current design for the sink logstore. So why can't we use the "unsynced" version here? What problems do the "synced" version want to solve?

Reply:
btw, why sink into table needs this new logstore, instead of the old one

Reply:
> I'm not very familiar with our current design for the sink logstore. So why can't we use the "unsynced" version here? What problems do the "synced" version want to solve?

Updated in PR description. Briefly:
1. Truncate the logstore before barrier gets committed, rather than asynchronously. This ensures exactly-once semantics.
2. Yield barriers downstream from the writer side. External sink does not need barriers so the desynced logstore does not have barriers yielded downstream.